Sources/RESP3/RESP3Error.swift

```diff
@@ -0,0 +1,79 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the RediStack open source project
+//
+// Copyright (c) 2023 RediStack project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of RediStack project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+
+/// This error is thrown if a RESP3 package could not be decoded.
+///
+/// If you see this error, there a two potential reasons this might happen:
+///
+///   1. The Swift RESP3 implementation is wrong
+///   2. You are contacting an untrusted backend
+///
+struct RESP3ParsingError: Error {
+    struct Code: Hashable, Sendable, CustomStringConvertible {
+        private enum Base {
+            case invalidLeadingByte
+            case invalidData
+            case tooDepplyNestedAggregatedTypes
+            case missingColonInVerbatimString
+            case canNotParseInteger
+            case canNotParseDouble
+            case canNotParseBigNumber
+        }
+
+        private let base: Base
```

Comment:
Could we add an index here? That way we can correlate the error back to a specific location in the received ByteBuffer. Definitely helps debugging should this ever trigger

Reply:
Sounds good, do you want me to make an issue for this?

Reply:
yeah go for it!